src/util/address.rs

```diff
@@ -406,6 +406,151 @@ impl Payload {
             } => script::Script::new_witness_program(version, prog)
         }
     }
+
+    /// Creates a pay to (compressed) public key hash payload from a public key
+    #[inline]
+    pub fn p2pkh(pk: &ecdsa::PublicKey) -> Payload {
+        let mut hash_engine = PubkeyHash::engine();
+        pk.write_into(&mut hash_engine)
+            .expect("engines don't error");
+        Payload::PubkeyHash(PubkeyHash::from_engine(hash_engine))
+    }
+
+    /// Creates a pay to script hash P2SH payload from a script
+    #[inline]
+    pub fn p2sh(script: &script::Script) -> Result<Payload, Error> {
+        if script.len() > MAX_SCRIPT_ELEMENT_SIZE {
+            return Err(Error::ExcessiveScriptSize);
+        }
+        Ok(Payload::ScriptHash(ScriptHash::hash(&script[..])))
+    }
+
+    /// Create a witness pay to public key payload from a public key
+    pub fn p2wpkh(pk: &ecdsa::PublicKey) -> Result<Payload, Error> {
+        if !pk.compressed {
+            return Err(Error::UncompressedPubkey);
+        }
+
+        let mut hash_engine = WPubkeyHash::engine();
+        pk.write_into(&mut hash_engine)
+            .expect("engines don't error");
+
+        Ok(Payload::WitnessProgram {
+            version: WitnessVersion::V0,
+            program: WPubkeyHash::from_engine(hash_engine)[..].to_vec(),
+        })
+    }
+
+    /// Create a pay to script payload that embeds a witness pay to public key
+    pub fn p2shwpkh(pk: &ecdsa::PublicKey) -> Result<Payload, Error> {
+        if !pk.compressed {
+            return Err(Error::UncompressedPubkey);
+        }
+
+        let mut hash_engine = WPubkeyHash::engine();
+        pk.write_into(&mut hash_engine)
+            .expect("engines don't error");
+
+        let builder = script::Builder::new()
+            .push_int(0)
+            .push_slice(&WPubkeyHash::from_engine(hash_engine)[..]);
+
+        Ok(Payload::ScriptHash(ScriptHash::hash(
+            builder.into_script().as_bytes(),
+        )))
+    }
+
+    /// Create a witness pay to script hash payload.
+    pub fn p2wsh(script: &script::Script) -> Payload {
+        Payload::WitnessProgram {
+            version: WitnessVersion::V0,
+            program: WScriptHash::hash(&script[..])[..].to_vec(),
+        }
+    }
+
+    /// Create a pay to script payload that embeds a witness pay to script hash address
+    pub fn p2shwsh(script: &script::Script) -> Payload {
+        let ws = script::Builder::new()
+            .push_int(0)
+            .push_slice(&WScriptHash::hash(&script[..])[..])
+            .into_script();
+
+        Payload::ScriptHash(ScriptHash::hash(&ws[..]))
+    }
+
+    /// Create a pay to taproot payload from untweaked key
+    pub fn p2tr<C: Verification>(
+        secp: Secp256k1<C>,
```

Comment:
This should be a reference to `Secp256k1<C>`

Reply:
The other part is being resolved here: #707

Reply:
Rebased this PR, should be good to go now!